sfxload.1
.TH sfxload 1 "January 22, 2003"
.LO 1
.SH NAME
sfxload, asfxload \- load a SoundFont file on the Emux WaveTable

.SH SYNOPSIS
.B sfxload
.RI [\| \-options \|] " fontfile"

.B asfxload
.RI [\| \-options \|] " fontfile"

.SH DESCRIPTION
.B sfxload
and
.B asfxload
are utility program sto transfer the sound wave
and instruments data in a SoundFont file to the AWE32 sound driver,
or to the Emux WaveTable of ALSA sbawe and emu10k1 drivers.
This program is necessary for playing MIDI samples via sequencer
programs supporting AWE driver.  There is no big difference between
\fBsfxload\fP and \fBasfxload\fP except for that \fBasfxload\fP is for
ALSA and \fBsfxload\fP is for OSS, respecitvely. The options to specify
devices are
different between them (see below).

Basically, sfxload behaves as two ways.
.in +1i
% sfxload fontfile
.br
% sfxload \-b1 fontfile
.in -1i

The first usage is to read SF2 (or SBK) file and transfer to the awe
driver.  In this case, the samples which were loaded on the driver are
replaced with the new one.
In the second case, sfxload reads the file and appends it to the
pre-loaded samples on the driver with specified bank number.  The old
samples remain in the driver.  The additional samples can be cleared
via \fB\-x\fP option (see below).

The sound files are searched through the path list.  The path list is
defined as built-in.  If the environment variable \fBSFBANKDIR\fP or the
command line option \fB\-P\fP is given, it replaces the default search list.
The file extension \fI.sf2\fP, and \fI.sbk\fP can be abbreviated.

.SH OPTIONS
.TP
.BI \-F,\ \-\-device= file " \fR(sfxload only)\fP"
Specify the device file to be used.
Default value is
.IR /dev/sequencer .
.TP
.BI \-D,\ \-\-index= number " \fR(sfxload only)\fP"
Specify the device index of AWE driver.
Negative number (e.g. \-1) means to probe the first AWE device automatically.
For selecting the other AWE cards, a positive index number must be given here.
Default value is \fB\-1\fP.
.TP
.BI \-D,\ \-\-hwdep= name " \fR(asfxload only)\fP"
Specify the hwdep name to be used.
As default, asfxload seeks until any Emux compatible hwdep device is found.
.TP
.BI \-i,\ \-\-clear "\fR[=\fPbool\fR]\fP"
Remove all samples before loading the fonts.
This is an explicit directive (see -b option).
If this option is specified alone without soundfont file
arguments, sfxload does onlay remove samples.
Either of \fBon\fP, \fBoff\fP, \fByes\fP, \fBno\fP, \fBtrue\fP,
or \fBfalse\fP can be specified as an optional argument.
.TP
.BI \-x,\ \-\-remove "\fR[=\fPbool\fR]\fP"
Remove the optional samples previouly loaded via \fB\-b\fP option.
Otherwise, all new samples are simply appended.
.TP	
.BI \-N,\ \-\-increment "\fR[=\fPbool\fR]\fP"
Do not clear samples even with the absence of \fB\-b\fP option.  However, this option is not exclusive with
.B \-x
option. If both options are specified, and the memory full error is
encountered during loading fonts, \fBsfxload\fP will try to remove
samples and load the fonts again.
.TP
.BI \-b,\ \-\-bank= number
Append the sound samples on the specified bank.
Without this option, all present samples in the driver are
removed before loading the new fonts unless \fB\-N\fP option is specified. Usually, this option is necessary to load user
bank, typically in bank one.
For example,
.in +1i
% sfxload synthgm.sbk
.br
% sfxload -b1 surprise.sf2
.br
.in -1i
.TP
.BI \-l,\ \-\-lock "\fR[=\fPbool\fR]\fP"
Lock the font.  The locked font is no longer removed via
remove option (\fB\-x\fP) even if it's loaded together with \fB\-b\fP option.
.TP
.BI \-C,\ \-\-compat "\fR[=\fPbool\fR]\fP"
Use the old (v0.4.2) parameter calculations.
.TP
.BI \-A,\ \-\-sense= sensitivity
(Only valid on compatible mode)
.br
Set sample attenuation sensitivity.
This option controls the sensitivity of initial attenuation parameter
of each instrument defined in SoundFont file.
In the program, each parameter is calculated from the value divided
by this number for the original value.
The number \fI1.0\fP means that the original initial attenuation parameters
would be used.
Generally, smaller number makes drum sounds louder.
(I think \fB"-A 2"\fP would be similar to Windows sounds.)
The default value is \fI10\fP.
Note that this option changes also the default attenuation automatically (see below).
.TP
.BI \-a,\ \-\-atten= attenuation
(Only valid on compatible mode)
.br
Set the default attenuation level.
This option controls the minimum attenuation of the sample.
The parameter is given in raw digit number for AWE driver,
that is, in 8/3 dB unit.
Since \fB\-\-sense\fP option changes the default attenuation automatically,
specify this option later from \fB\-\-sense\fP option.
The default value is \fI32\fP.
.TP
.BI \-d,\ \-\-decay= scale
(Only valid on compatible mode)
.br
Set the scale of envelope decay time.  Default value is \fI50.0\fP.
Sounds decay fast when larger number is set.
The ver.0.3 sfxload uses \fI54.8\fP.  If you want to keep the same
sound, use this number instead.
.TP
.BI \-M,\ \-\-memory "\fR[=\fPbool\fR]\fP"
Display the left memory size in DRAM on the AWE32 card.
.TP
.BI \-c,\ \-\-chorus= percent
Specify the effects of chorus.
The value is in percent, from 0 to 100.  The default is unspecified.
This value may be overwritten by MIDI control messages.
.TP
.BI \-r,\ \-\-reverb= percent
Specify the effects of reverb.
The value is in percent, from 0 to 100.  The default is unspecified.
This value may be overwritten by MIDI control messages.
.TP
.BI \-B,\ \-\-addblank "\fR[=\fPbool\fR]\fP"
Add 48 size of blank loop on each sample data.  Usually, this
option is not necessary.  Most of soundfont files are designed 
well for enough blank loops for each sample.
.TP
.BI \-L,\ \-\-extract= source \fR[:\fP map \fR]\fP
Extract and load only the specified preset(s).
This option is usually employed by drvmidi.
The preset is given as same as in virtual bank file.
.TP
.BI \-v,\ \-\-verbose "\fR[=\fPlevel\fR]\fP"
Increase or set the verbosity level.
.TP
.BI \-q,\ \-\-quiet
Don't show error messages, equivalen with \-\-verbose=0.
.TP
.BI \-V,\ \-\-volume= percent
Specify the total volume of sounds, provided in percent.
The default volume is \fI70%\fP.
.TP
.BI \-P,\ \-\-path= path1 : path2 : pathN
Specify the search path list.
The sound files are searched from
\fIpath1\fP, \fIpath2\fP, and so on.
This overrides both the system path and environment variable
\fBSFBANKDIR\fP.

.SH "VIRTUAL BANK FILE"
The virtual bank file is a list of presets treated as one soundfont
file.  The syntax of virtual bank is as follows:
.in +1i
	# comments
.br
	source:map[:soundfont [preset-name]
.br
	source:map[:soundfont [preset-name]
.br
	...
.in -1i

The first and second items are the source and mapped presets,
respectively.  The former is the existing preset in the soundfont,
and the latter is the actual preset loaded on the sound driver.
The preset is described by the following three values,
.in +1i
	preset/bank/keynote
.in -1i
If bank and keynote are omitted, bank 0 and keynote \-1 (meaning to
search all keys) are applied.

The third item is the name of soundfont file.  The file is searched
from the prescribed search-path.  The remaining arguments are
ignored in \fBsfxload\fP.

If the soundfont name is omitted, sfxload loads it as preset
mapping.  It just behaves like a symbolic link of file opposing to
copying of the file.  Any sample data is not referred at this time,
but searched first when the note is played.

A couple of special commands can be used together with the virtual
presets above.
\fBdefault\fP
command is used to specify the default
soundfont file.  Any other presets which are not defined in the
virtual preset lists are loaded from this default font.  For
example, in the following virtual bank, \fI2mbgmgs.sf2\fP is used except
for standard drumsets which employs \fIdrum.sf2\fP:
.in +1i 
	0/128:0/128:drum.sf2
.br
	default 2mbgmgs.sf2
.in -1i

Another special command is
\fBinclude\fP
command.  This simply includes
another virtual bank file under the current position. For example,
.in +1i
	default standard.sf2
.br
	0/128:0/128:drum.sf2
.br
	include xgsfx.bnk
.in -1i

.SH "SYSTEM RESROUCE FILE"

The default option arguments can be stored in the system resource file.
There are two files loaded as default.
.in +1i
.I $HOME/.sfxloadrc
.br
.I /etc/sfxloadrc
.in -1i

The syntax is as follows:
.in +1i
fontname \-options..
.in -1i
The first argument is soundfont file name for each option.
The remaining arguments are identical with command line options.
The font name
\fIdefault\fP
is used for default options for all sound
fonts.  The default options are overridden by specific options or
command line options.

For example, to set default chorus level 20 except for synthgm font,
you can write a resource file \fI~/.sfxloadrc\fP like that:
.in +1i
default	\-\-chorus=20
.br
synthgm \-\-chorus=0
.in -1i


.SH ENVIRONMENT
.TP
.B SFBANKDIR
Search path for sound files.  The current directory is always searched
at first.

.SH "SEE ALSO"
.BR drvmidi (1)

.SH COPYRIGHT
Copyright (C) 1996-2003 Takashi Iwai.
.P
The AWE32 driver and utilties are free software; you can redistribute them
and/or modify them under the terms of the \fIGNU General Public
License\fP as published by the Free Software Foundation; either
version 2 of the License, or (at your option) any later version.


ChangeLog
ver.0.5.1
	- include awe_voice.h into the package (to fix build with 2.6.22
	  kernel)
	- added udev-related files to etc subdirectory
	
ver.0.5.0d
	- fixed compilation with gcc-2.x.
	- added -q (--quiet) option to [a]sfxload.
	  verbosity is set to 1 as default.
	  some debug messages appears with level 2.
	- fixed the pitch correction type to signed.
	  this will fix some weird detuned tones.

ver.0.5.0c
	- fixed the option parsing.  -b option works again now.

ver.0.5.0b
	- patches by N. Vignot
	  - fixed sf2text outputs for non-printable letters
	  - fixed fskip for pipe
	  - added .bz2 extention for open

ver.0.5.0a
	- patches by Philipp Matthias Hahn <pmhahn@titan.lahn.de>
	  - fixed compile warnings
	  - fixed the wrong options in aweset
	  - fixed manpage

ver.0.5.0
	- Support ALSA emux hwdep interface.  asfxload is added.
	  The hardware access operators are packed as AWEOps struct,
	  the awelib no longer accesses the DSP i/o directly.
	- Use GNU auto-tools instead of Imake.
	- Moved docs and some files in the root directory for GNU style.
	- Fixed some buffer-overflows.
	- Build awelib static.
	- Added awe_ prefix to some awelib functions to avoid name
	  conflictions.
	- Changed the default SoundFont path.
	
ver.0.4.4
	- Fixed Makefiles
	- Fixed big-endian handling

ver.0.4.3c
	- Add --device and --index options
	- RPM spec file is included

ver.0.4.3b
	- Change directory contents
	- Add sapmles Makefiles for shared-library
ver.0.4.3a
	- Minor bug fix
ver.0.4.3
	- Official relase

ver.0.4.3p4
	- Add -i and -N options

ver.0.4.3p3
	- Load via sample sharing mode (for awedrv-0.4.3p4)
	- Fix a bug to miss some fonts in dynamic loading
	- Fix a bug in path.c
	- Change library version to 0.4.3

ver.0.4.3p2
	- Fix a bug to garbage purge_list (in parsesf.c)
	- New program: aweset

ver.0.4.3p1
	- Change parameter calculation as well as DOS drivers
	- Add --compat option to keep the compatibility with older versions
	- Avoid to load a part of layers when out of memory

ver.0.4.2
	- Add setfx command.
	- Add lock option.
	- Support for virtual bank file.
	- Support for preset mapping
	- Merge global variables to one option structure.
	- Change from Makefile to Imakefile
	- Create AWElib for easy use of patch loading
	- Fix a getopt bug.
	- Fix a bug in calculation of envelope hold time
	- Fix a bug in calculation of coarse loop addresses
	- Sfxload tells if the file is not found
	- Add keynote option in gusload
	- Change sfxtest arguments

ver.0.4.1	development versions

ver.0.4.0
	TOTALLY REWRITEN.
	- Remove -i option: always remove samples if not speicifed -b
	  option.
	- Remove other obsolte options, -I, -s, -m.
	  (though they're remained for comaptibility.)
	- Add attenuation control options, -A and -a.
	- Add memory display option, -M.
	- Do not support SFX file in sfxload utility.
	- Read & write S-expression style text file.
	- Add partial loading
	- Use default rc file
	- Support long options


sfxload.c
/*================================================================
 * sfxload -- load soundfont info onto awe sound driver
 *
 * Copyright (C) 1996-2000 Takashi Iwai
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *================================================================*/

#include <stdio.h>
#include <ctype.h>
#include <unistd.h>
#include <string.h>
#include <errno.h>
#include <stdlib.h>
#include <sys/fcntl.h>
#include <util.h>
#ifdef BUILD_ASFXLOAD
#include <alsa/asoundlib.h>
#endif
#include <awebank.h>
#include <sfopts.h>
#include <awe_version.h>
#include "seq.h"

#ifdef BUILD_ASFXLOAD
#define PROGNAME "asfxload"
#else
#define PROGNAME "sfxload"
#endif

static char *get_fontname(int argc, char **argv);
static int parse_options(int argc, char **argv);
static void add_part_list(char *arg);

extern int awe_verbose;

static AWEOps load_ops = {
	seq_load_patch,
	seq_mem_avail,
	seq_reset_samples,
	seq_remove_samples,
	seq_zero_atten
};



/*----------------------------------------------------------------
 * print usage and exit
 *----------------------------------------------------------------*/

static void usage()
{
	fputs(
#ifdef BUILD_ASFXLOAD
	      "asfxload -- load SoundFont on ALSA Emux WaveTable\n"
#else
	      "sfxload -- load SoundFont on OSS AWE32 sound driver\n"
#endif
	      VERSION_NOTE
	      "usage:	" PROGNAME " [-options] [soundfont[.sf2|.sbk|.bnk]]\n"
	      "\n"
	      " options:\n"
#ifdef BUILD_ASFXLOAD
	      " -D, --hwdep=name        specify the hwdep name\n"
#else
	      " -F, --device=file        specify the device file\n"
	      " -D, --index=number       specify the device index (-1=autoprobe)\n"
#endif
	      " -i, --clear[=bool]       clear all samples\n"
	      " -x, --remove[=bool]      remove additional samples\n"
	      " -N, --increment[=bool]   incremental loading\n"
	      " -b, --bank=number        append font to the specified bank\n"
	      " -l, --lock[=bool]        lock the loading fonts\n"
	      " -v, --verbose[=int]      set verbosity level\n"
	      " -q, --quiet              don't print error messages\n"
	      " -C, --compat[=bool]      use v0.4.2 compatible sounds\n",
	      stderr);
	fprintf(stderr, " -A, --sense=digit        (compat) set attenuation sensitivity (default=%g)\n", awe_option.atten_sense);
	fprintf(stderr, " -a, --atten=digit        (compat) set default attenuattion (default=%d)\n", awe_option.default_atten);
	fprintf(stderr, " -d, --decay=scale        (compat) set decay time scale (default=%g)\n", awe_option.decay_sense);
	fputs(" -M, --memory[=bool]      display available memory on DRAM\n"
	      " -B, --addblank[=bool]    add 12 words blank loop on each sample\n"
	      " -c, --chorus=percent     set chorus effect (0-100)\n"
	      " -r, --reverb=percent     set reverb effect (0-100)\n",
	      stderr);
	fprintf(stderr, " -V, --volume=percent     set total volume (0-100) (default=%d)\n", awe_option.default_volume);
	fputs(" -L, --extract=preset/bank/note\n"
	      "                          do partial loading\n"
	      " -P, --path=dir           set SoundFont file search path\n",
	      stderr);
	if (awe_option.search_path)
		fprintf(stderr, "   system default path is %s\n", awe_option.search_path);
	exit(1);
}


static int sample_mode, remove_samples, dispmem, lock_sf;
enum { CLEAR_SAMPLE, INCREMENT_SAMPLE, ADD_SAMPLE };
#ifdef BUILD_ASFXLOAD
static char *hwdep_name = NULL;
#else
static char *seq_devname = NULL;
static int seq_devidx = -1;
#endif

static LoadList *part_list = NULL;

int main(int argc, char **argv)
{
	char *sffile;
	int rc;

	awe_init_option();

	sample_mode = ADD_SAMPLE;
	remove_samples = FALSE;
	dispmem = FALSE;
	lock_sf = -1;
	awe_verbose = 1;
	part_list = NULL;

	awe_read_option_file(NULL);
	if ((sffile = get_fontname(argc, argv)) != NULL) {
		awe_read_option_file(sffile);
	}

	parse_options(argc, argv);

	DEBUG(1,fprintf(stderr, "default bank = %d\n", awe_option.default_bank));
	DEBUG(1,fprintf(stderr, "default chorus = %d\n", awe_option.default_chorus));
	DEBUG(1,fprintf(stderr, "default reverb = %d\n", awe_option.default_reverb));
	if (awe_option.compatible) {
		DEBUG(1,fprintf(stderr, "v0.4.2-compatible mode\n"));
		DEBUG(1,fprintf(stderr, "minimum attenuation = %d\n", awe_option.default_atten));
		DEBUG(1,fprintf(stderr, "attenuation sense = %g\n", awe_option.atten_sense));
		DEBUG(1,fprintf(stderr, "decay sense = %g\n", awe_option.decay_sense));
	} else {
		DEBUG(1,fprintf(stderr, "use new calculation\n"));
	}

	/*----------------------------------------------------------------*/
#ifdef BUILD_ASFXLOAD
	seq_alsa_init(hwdep_name);
#else
	/* reset samples if necessary */
	seq_init(seq_devname, seq_devidx);
#endif

	/* clear or remove samples */

	if (sample_mode == ADD_SAMPLE && awe_option.default_bank < 0 && sffile)
		sample_mode = CLEAR_SAMPLE;

	if (sample_mode == CLEAR_SAMPLE)
		seq_reset_samples();
	else if (sample_mode != INCREMENT_SAMPLE && remove_samples)
		seq_remove_samples();

	if (sffile == NULL && dispmem) {
		if (dispmem)
			printf("DRAM memory left = %d kB\n", seq_mem_avail()/1024);
	}
	if (sffile == NULL) {
#ifdef BUILD_ASFXLOAD
		seq_alsa_end();
#else
		seq_end();
#endif
		return 0;
	}

	/* if lock option is not specified, lockinig depends on bank mode */
	if (lock_sf < 0) {
		if (awe_option.default_bank < 0)
			lock_sf = TRUE;
		else
			lock_sf = FALSE;
	}

	rc = awe_load_bank(&load_ops, sffile, part_list, lock_sf);
	if (sample_mode == INCREMENT_SAMPLE && remove_samples) {
		if (rc == AWE_RET_NOMEM) {
			seq_remove_samples();
			rc = awe_load_bank(&load_ops, sffile, part_list, lock_sf);
		}
	}

	if (rc == AWE_RET_OK && dispmem)
		printf("DRAM memory left = %d kB\n", seq_mem_avail()/1024);

#ifdef BUILD_ASFXLOAD
		seq_alsa_end();
#else
	seq_end();
#endif
	if (rc == AWE_RET_NOMEM) {
		if (awe_verbose)
			fprintf(stderr, "sfxload: no memory left\n");
		return 1;
	} else if (rc == AWE_RET_NOT_FOUND) {
		fprintf(stderr, "sfxload: can't find font file %s\n", sffile);
		return 1;
	} else if (rc == AWE_RET_ERR) {
		if (awe_verbose)
			fprintf(stderr, "sfxload: stopped by error\n");
	}

	return 0;
}


/*----------------------------------------------------------------
 * long options
 *----------------------------------------------------------------*/

static struct option long_options[] = {
	{"memory", 2, 0, 'M'},
	{"remove", 2, 0, 'x'},
	{"increment", 2, 0, 'N'},
	{"clear", 2, 0, 'i'},
	{"verbose", 2, 0, 'v'},
	{"quiet", 0, 0, 'q'},
	{"extract", 1, 0, 'L'},
	{"lock", 2, 0, 'l'},
#ifdef BUILD_ASFXLOAD
	{"hwdep", 1, 0, 'D'},
#else
	{"device", 1, 0, 'F'},
	{"index", 1, 0, 'D'},
#endif
	{0, 0, 0, 0},
};
static int option_index;

#ifdef BUILD_ASFXLOAD
#define OPTION_FLAGS	"MxNivqL:lD:"
#else
#define OPTION_FLAGS	"MxNivqL:lF:D:"
#endif

int awe_get_argument(int argc, char **argv, char *optstr, struct option *args)
{
	int c, dummy;
	int optind_saved;
	optind_saved = optind;
	optind = 0;
	while ((c = getopt_long(argc, argv, optstr, args, &dummy)) != -1)
		;
	c = optind;
	optind = optind_saved;
	return c;
}

static char *get_fontname(int argc, char **argv)
{
	int rc;
	rc = awe_get_argument(argc, argv, OPTION_FLAGS AWE_BASE_OPTIONS, long_options);
	if (rc < argc)
		return argv[rc];
	return NULL;
}

#define set_bool()	(optarg ? bool_val(optarg) : TRUE)

static int parse_options(int argc, char **argv)
{
	int c;
	while ((c = awe_parse_options(argc, argv, OPTION_FLAGS, long_options, &option_index)) != -1) {
		if (c == 0)
			continue;
		switch (c) {
		case 'x':
			remove_samples = set_bool();
			break;
		case 'N':
			if (set_bool())
				sample_mode = INCREMENT_SAMPLE;
			break;
		case 'i':
			if (set_bool())
				sample_mode = CLEAR_SAMPLE;
			break;
		case 'M':
			dispmem = set_bool();
			break;
		case 'v':
			if (optarg)
				awe_verbose = atoi(optarg);
			else
				awe_verbose++;
			break;
		case 'q':
			awe_verbose = 0;
			break;
		case 'L':
			add_part_list(optarg);
			break;
		case 'l':
			lock_sf = set_bool();
			break;

#ifdef BUILD_ASFXLOAD
		case 'D':
			hwdep_name = optarg;
			break;
#else
		case 'F':
			seq_devname = optarg;
			break;
		case 'D':
			seq_devidx = atoi(optarg);
			break;
#endif
		case 'm':
		case 's':
		case 'I':
			break;

		default:
			usage();
			exit(1);
		}
	}
	return -1;
}

/* make a preset list from comand line options */

static void add_part_list(char *arg)
{
	char tmp[100];
	SFPatchRec pat, map;
	if (strlen(arg) > sizeof(tmp)-1) {
		fprintf(stderr, "sfxload: illegal argument %s\n", arg);
		return;
	}
	strcpy(tmp, arg);
	if (awe_parse_loadlist(tmp, &pat, &map, NULL))
		part_list = awe_add_loadlist(part_list, &pat, &map);
}


